proposals/type_inference.md

```diff
@@ -0,0 +1,163 @@
+
+### Program 4 
+
+	x = (1 + 3) * 2
+	y = x * 2
+	z = x + y 
+
+From the types of x and y we are able to infer the type of z from the right hand side. 
+
+### Program 5
+
+	class List(object):
+	  def sum(self : List) -> int:
+	    return 1 // 0 # Intentional error! Make sure we implement in subclasses
+
+	class Empty(List):
+	  def sum(self : List) -> int:
+	    return 0
+
+	class Link(List):
+	  val : int = 0
+	  next : List = None
+	  def sum(self : Link) -> int:
+	    return self.val + self.next.sum()
+	  
+	  def new(self : Link, val : int, next : List) -> Link:
+	    self.val = val
+	    self.next = next
+	    return self
+
+	l: List = Link().new(5, Link().new(1, Empty())) 
+	print(l.sum())
+	print(l)
+
+We expect for l to be type List since the user explicitly annotated it. If that annotation were missing then the type of l would be inferred as Link. 
+
+### Program 6
+
+	def g(z: int): 
+	  def f(y): 
+	    return y + 1
+	  x = f(1)
+
+From the type of the argument of f we can infer the return type of f, from which we can infer the type of x
+
+### Program 7
+
+	def g(z: int): 
+	  z = z + 1 
+	  return z
+
+Here since the type of z is known we can infer the return type of g from its body. 
+
+### Program 8
+
+	T = TypeVar('T')      
+
+	def first(l: Sequence[T]) -> T:
+	  return l[0]
+
+Users can define type variables for generic syntax. 
+
+### Program 9 
+
+	class A:
+	   x: int = 5
+	   y: int = 3
+
+	class B:
+	  x: str = ‘5’
+	  y: str = ‘3’
+
+	def f(obj): 
+	  return obj.x + obj.y  #this needs some kind of defined behavior
+
+We’re still considering how to handle programs that like this in a way that is compatible with REPL. Options that we have thought about so far
+- Error message indicating that type annotations are needed. 
+- Determine types based on call sites in code (will break in REPL). 
+
+### Program 10
+
+	# Before: 
+	T = TypeVar('T')
+	def f(x: T) -> T: ...
+	
+	# Later in REPL: 
+	f(1)                   
+	f('a')   
+
+Infer the type of the function based on the types of the arguments passed at REPL so the above should be valid for sensible implementations of f. 
+
+## Testing
+
+Our testing strategy will be driven by unit tests at the following levels*:
+- Check that a set of reasonable expressions types are inferred correctly 
+	- e.g. 1, True, None, constructor calls, and variable usages   
+- Check AST forms produced after type inference and type checking
+	- Ensure the inferred types are correct
+	- Ensure that users can declare variable as a super type to override the type that would have been inferred successfully
+	- Ensure generic types exist and their usage is valid 
+- Check that programs that use type inference produce correct outputs
+- Check that reasonable errors are produced when more type information is required 
+
+*We may also consider exploring property-based testing if time permits. 
+
+## Ast Forms
+
+- For Local Type Inference: No changes should be needed
```

Comment:
Is this true? Do you know if you need to change anything about the `VarInit` sequences in the prelude of function bodies?

Reply:
I think this was written under the assumption that we may not necessarily start from this particular compiler implementation. It seems like you want us to use this compiler, so should we adjust our proposal under this assumption?